flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -861,23 +865,32 @@ private TableResult executeOperation(Operation operation) {
 					catalogManager.getTable(describeTableOperation.getSqlIdentifier());
 			if (result.isPresent()) {
 				TableSchema schema = result.get().getTable().getSchema();
-				String[][] rows = Stream.concat(
+				Map<String, String> fieldToWatermark =
+					schema.getWatermarkSpecs()
+						.stream()
+							.collect(Collectors.toMap(WatermarkSpec::getRowtimeAttribute, WatermarkSpec::getWatermarkExpr));
+
+				Map<String, String> fieldToPrimaryKey = new HashMap<>();
+				schema.getPrimaryKey().ifPresent((p) -> {
+					List<String> columns = p.getColumns();
+					columns.forEach((c) -> fieldToPrimaryKey.put(c, String.format("PRI(%s)", String.join(",", columns))));
+				});
+
+				String[][] rows =
 					schema.getTableColumns()
 						.stream()
-							.map((c) -> new String[]{
+							.map((c) -> {
+								LogicalType logicalType = c.getType().getLogicalType();
+								return new String[]{
 								c.getName(),
-								c.getType().getLogicalType().toString(),
-								c.getExpr().orElse("(NULL)")}),
-					schema.getWatermarkSpecs()
-						.stream()
-							.map((w) -> new String[]{
-								"WATERMARK",
-								"(NULL)",
-								w.getWatermarkExpr()
-							})
-				).toArray(String[][]::new);
-
-				return buildShowResult(new String[]{"name", "type", "expr"}, rows);
+								StringUtils.removeEnd(logicalType.toString(), " NOT NULL"),
+								logicalType.isNullable() ? "true" : "false",
+								fieldToPrimaryKey.getOrDefault(c.getName(), "(NULL)"),
```

Comment:
we should use `null` instead of `(NULL)`, `(NULL)` just a display form defined in `PrintUtils`.